packages/babel-parser/src/parser/scope.js

```diff
@@ -0,0 +1,152 @@
+// @flow
+import UtilParser from "./util";
+import {
+  SCOPE_ARROW,
+  SCOPE_ASYNC,
+  SCOPE_DIRECT_SUPER,
+  SCOPE_FUNCTION,
+  SCOPE_GENERATOR,
+  SCOPE_SIMPLE_CATCH,
+  SCOPE_SUPER,
+  SCOPE_TOP,
+  SCOPE_VAR,
+  BIND_SIMPLE_CATCH,
+  BIND_LEXICAL,
+  BIND_FUNCTION,
+  type ScopeFlags,
+  type BindingTypes,
+  SCOPE_CLASS,
+} from "../util/scopeflags";
+
+// Start an AST node, attaching a start offset.
+class Scope {
+  flags: ScopeFlags;
+  // A list of var-declared names in the current lexical scope
+  var: string[] = [];
+  // A list of lexically-declared names in the current lexical scope
+  lexical: string[] = [];
+  // A list of lexically-declared FunctionDeclaration names in the current lexical scope
+  functions: string[] = [];
+
+  constructor(flags: ScopeFlags) {
+    this.flags = flags;
+  }
+}
+
+// The functions in this module keep track of declared variables in the
+// current scope in order to detect duplicate variable names.
+export default class ScopeParser extends UtilParser {
```

Comment:
I changed it to be a separate class, which works and might be nice to have like this if we would want to reuse this parts in traverse for example.

But when looking into flow I noticed another problem: extending the scope tracking is not as easy possible with the external class. So for typescript we probably would want to add another binding group for types. Flow has some custom rules too when it comes to `declare class|var|function`. I'm not sure if we actually want to support extending the scope tracking but it might be nice.